modulus/models/gnn_layers/distributed_graph.py

```diff
@@ -117,94 +131,111 @@ def partition_graph_nodewise(
+                f"Aborting partitioning, rank {rank} has 0 source nodes to work on."
+            )
+
+        # create mapping of global IDs to local IDs
+        # as each rank is expected to operate on disting global IDs, this is expected
+        # to not cause any data races
+        ids = src_nodes_in_each_partition[rank]
+        mapping_global_src_ids_to_local_ids[ids] = torch.arange(
+            ids.numel(),
+            dtype=mapping_global_src_ids_to_local_ids.dtype,
+            device=mapping_global_src_ids_to_local_ids.device,
+        )
+
+    graph_partition.num_src_nodes_in_each_partition = num_src_nodes_in_each_partition
+    graph_partition.num_dst_nodes_in_each_partition = num_dst_nodes_in_each_partition
 
+    # create local graph structures
     for rank in range(partition_size):
-        offset_start = dst_offsets_in_partition[rank]
-        offset_end = dst_offsets_in_partition[rank + 1]
-        offsets = global_offsets[offset_start : offset_end + 1].detach().clone()
-        partition_indices = global_indices[offsets[0] : offsets[-1]].detach().clone()
-        offsets -= offsets[0].item()
+        offset_start = global_offsets[dst_nodes_in_each_partition[rank]].view(-1, 1)
+        offset_end = global_offsets[dst_nodes_in_each_partition[rank] + 1].view(-1, 1)
+        degree = offset_end - offset_start
+        local_offsets = degree.view(-1).cumsum(dim=0)
+        local_offsets = torch.cat(
+            [
+                torch.Tensor([0]).to(
+                    dtype=local_offsets.dtype, device=local_offsets.device
+                ),
+                local_offsets,
+            ]
+        )
+
+        # create boolean mask to find contigouus sections of global_indices
+        # which are taken care of current rank without using loops
+        tmp = torch.arange(
+            global_indices.numel(),
+            dtype=global_indices.dtype,
+            device=global_indices.device,
+        )
+        mask = (offset_start <= tmp) & (tmp < offset_end)
+        mask = torch.any(mask, dim=0)
 
-        global_src_ids_per_rank, inverse_mapping = partition_indices.unique(
+        partition_indices = global_indices[mask].detach().clone()
+        global_src_ids_on_rank, inverse_mapping = partition_indices.unique(
             sorted=True, return_inverse=True
         )
-        local_src_ids_per_rank = torch.arange(
+        local_src_ids_on_rank = torch.arange(
             0,
-            global_src_ids_per_rank.size(0),
-            dtype=offsets.dtype,
-            device=offsets.device,
-        )
-        global_src_ids_to_gpu = global_src_ids_per_rank // src_nodes_in_partition
-        remote_src_ids_per_rank = (
-            global_src_ids_per_rank - global_src_ids_to_gpu * src_nodes_in_partition
+            global_src_ids_on_rank.size(0),
+            dtype=local_offsets.dtype,
+            device=local_offsets.device,
         )
+        remote_src_ids_on_rank = mapping_global_src_ids_to_local_ids[
```

Comment:
Oh, I see the distinction between local and remote src nodes. Might be good to clarify this in the docstring